Source/WebCore/html/shadow/SpatialImageControls.cpp

```diff
@@ -0,0 +1,214 @@
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "SpatialImageControls.h"
+
+#include "ElementInlines.h"
+#include "ElementRareData.h"
+#include "Event.h"
+#include "EventLoop.h"
+#include "HTMLButtonElement.h"
+#include "HTMLDivElement.h"
+#include "HTMLImageElement.h"
+#include "HTMLNames.h"
+#include "HTMLStyleElement.h"
+#include "MouseEvent.h"
+#include "RenderImage.h"
+#include "ShadowRoot.h"
+#include "TreeScopeInlines.h"
+#include "UserAgentStyleSheets.h"
+#include <wtf/text/AtomString.h>
+
+namespace WebCore {
+namespace SpatialImageControls {
+
+#if ENABLE(SPATIAL_IMAGE_CONTROLS)
+
+static const AtomString& spatialImageControlsElementIdentifier()
+{
+    static MainThreadNeverDestroyed<const AtomString> identifier("spatial-image-controls"_s);
+    return identifier;
+}
+
+static const AtomString& spatialImageControlsButtonIdentifier()
+{
+    static MainThreadNeverDestroyed<const AtomString> identifier("spatial-image-controls-button"_s);
+    return identifier;
+}
+
+bool hasSpatialImageControls(const HTMLElement& element)
+{
+    RefPtr shadowRoot = element.shadowRoot();
+    if (!shadowRoot || shadowRoot->mode() != ShadowRootMode::UserAgent)
+        return false;
+
+    return shadowRoot->hasElementWithId(spatialImageControlsElementIdentifier());
+}
+
+static RefPtr<HTMLElement> spatialImageControlsHost(const Node& node)
+{
+    RefPtr host = dynamicDowncast<HTMLElement>(node.shadowHost());
+    if (!host)
+        return nullptr;
+
+    return hasSpatialImageControls(*host) ? host : nullptr;
+}
+
+bool isSpatialImageControlsButtonElement(const Element& element)
+{
+    return spatialImageControlsHost(element) && element.getIdAttribute() == spatialImageControlsButtonIdentifier();
+}
+
+bool shouldHaveSpatialControls(HTMLImageElement& element)
+{
+    if (!element.document().settings().spatialImageControlsEnabled())
+        return false;
+
+    bool hasSpatialcontrolsAttribute = element.hasAttributeWithoutSynchronization(HTMLNames::controlsAttr);
+
+    auto* cachedImage = element.cachedImage();
+    if (!cachedImage)
+        return false;
+
+    auto* image = cachedImage->image();
+    if (!image)
+        return false;
+
+    return hasSpatialcontrolsAttribute && image->isSpatial();
+}
+
+void ensureSpatialControls(HTMLImageElement& imageElement)
+{
+    if (!shouldHaveSpatialControls(imageElement) || hasSpatialImageControls(imageElement))
+        return;
+
+    imageElement.protectedDocument()->checkedEventLoop()->queueTask(TaskSource::InternalAsyncTask, [weakElement = WeakPtr { imageElement }] {
+        RefPtr element = weakElement.get();
+        RefPtr shadowRoot = element->shadowRoot();
```

Comment:
We need to check the nullity of element. Element could be dead between the time this callback is scheduled and when it's called so add:
```cpp
if (!element)
    return;
```